### PR TITLE
DIS2-975 kill pgid

### PR DESCRIPTION
**Pull request by johnhuangclever**

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 machine:
   post:
   - cd $HOME && git clone --depth 1 -v git@github.com:clever/ci-scripts.git && cd ci-scripts && git show --oneline -s
-  - $HOME/ci-scripts/circleci/golang-install 1.8
+  - $HOME/ci-scripts/circleci/golang-install 1.9
```
Comment (natebrennand) [inline circle.yml]: why upgrade?

File: gearcmd/worker.go [hunk 2]
```diff
@@ -302,7 +302,14 @@ func stopProcess(p *os.Process, gracePeriod time.Duration) error {
 		}
 		time.Sleep(gracePeriod)
 	}
-	if err := syscall.Kill(p.Pid, syscall.SIGKILL); err != nil {
+	// kill entire group of process spawned by our cmd.Process
+	pgid, err := syscall.Getpgid(p.Pid)
+	if err != nil {
+		return fmt.Errorf("unable to get pgid, error: %s", err)
```
Comment (natebrennand) [inline gearcmd/worker.go]: I think we should not fail here, why not log and then still kill the base process?

This is how I had started to do it:
```diff
diff --git a/gearcmd/worker.go b/gearcmd/worker.go
index eaccec5..e48c57c 100644
--- a/gearcmd/worker.go
+++ b/gearcmd/worker.go
@@ -302,7 +302,14 @@ func stopProcess(p *os.Process, gracePeriod time.Duration) error {
                }
                time.Sleep(gracePeriod)
        }
-       if err := syscall.Kill(p.Pid, syscall.SIGKILL); err != nil {
+
+       targetID := p.Pid
+       pgid, err := syscall.Getpgid(p.Pid)
+       if err == nil {
+               targetID = -pgid
+       }
+
+       if err := syscall.Kill(targetID, syscall.SIGKILL); err != nil {
                // The graceful shutdown may have completed, so we don't error in that case.
                if gracePeriod == 0 || !strings.Contains(err.Error(), "no such process") {
                        return fmt.Errorf("unable to send SIGKILL, error: %s", err)
```

File: gearcmd/worker.go [hunk 3]
```diff
@@ -302,7 +302,14 @@ func stopProcess(p *os.Process, gracePeriod time.Duration) error {
...
+	}
+	lg.InfoD("killing-pgid", logger.M{"pid": p.Pid})
```
Comment (natebrennand) [inline gearcmd/worker.go]: this seems out of date

File: gearcmd/worker.go [hunk 4]
```diff
@@ -302,7 +302,14 @@ func stopProcess(p *os.Process, gracePeriod time.Duration) error {
...
+	// minus sign required to kill PGIDs
```
Comment (natebrennand) [inline gearcmd/worker.go]: Can we flesh this out a bit more?
I think a link to https://linux.die.net/man/2/kill would be useful

File: circle.yml (hunk 1 above)
Comment (johnhuangclever) [reply circle.yml]: That's what's on my computer. Is there a reason we shouldn't?

File: gearcmd/worker.go (hunk 2 above)
Comment (johnhuangclever) [reply gearcmd/worker.go]: 👍

File: circle.yml (hunk 1 above)
Comment (natebrennand) [reply circle.yml]: Since we run 1.8 everywhere else in production. 1.9 is under 1 month old so it feels a little pre-mature to move over.

**Comment by natebrennand** [comment]
Question on testing is still standing

File: circle.yml (hunk 1 above)
Comment (johnhuangclever) [reply circle.yml]: k i'll keep it as is

**Comment by johnhuangclever** [comment]
Tested locally using some scripts similar to the original gaprov-worker script:

```
#!/bin/bash
# test.sh

echo "starting"
node_modules/.bin/coffee test_script.coffee $1
echo "okay"
sleep 3m
echo "done"
```

```
# test_script.coffee
console.log process.argv

setTimeout ->
  console.log("done")
, 600000000
```

`build/gearcmd-v0.8.7-darwin-amd64/gearcmd -parseargs=false -name john -cmd ./test.sh -cmdtimeout 5s -host=clever-dev--gearmand.int.clever.com -port=4730`

master build leaves coffeescript processes around:
```
johnhuang:~/go/src/github.com/Clever/signalfx-metadata-sync$ ps -o pgid -ef | less -i
johnhuang:~/go/src/github.com/Clever/signalfx-metadata-sync$ ps -o pgid -ef | grep coffee
PGID   UID   PID  PPID   C STIME   TTY           TIME CMD
26426   501 26427     1   0  6:44PM ttys005    0:00.22 node node_modules/.bin/coffee test_script.coffee {"district_id":"55b0160f4307c901000001c6"}
26434   501 26435     1   0  6:44PM ttys005    0:00.25 node node_modules/.bin/coffee test_script.coffee {"district_id":"55b0160f4307c901000001c6"}
```
This branch cleans them all up

**Pull request merged**